### ISSUE TITLE
Unable to ez pz install on windows

### ISSUE THREAD

**Issue by NoeJuza** [invoke-ai/InvokeAI#748]
**Describe your environment**
- GPU: CUDA
- VRAM: 8GB
- CPU arch: x86
- OS: Windows
- Python: other (explain)
- Branch: main
- Commit: a1739a73b48bfe98b6abcb67f5a0197a9ad270e0

**Describe the bug**
I was following the documentation [here] (https://github.com/invoke-ai/InvokeAI/wiki/Easy-peasy-Windows-install) and got stuck on step 7.iv because the creation of the virtual python environment outputs with an error see [this pastebin for more info] (https://pastebin.com/wzE1HVLu) (full log of the error).

**To Reproduce**
1. Follow the procedure [here] (https://github.com/invoke-ai/InvokeAI/wiki/Easy-peasy-Windows-install) up to point 7.iv where you create the virtual environment.
2. paste the command and modify the requirements source to the "win" requirements because the -lin-win-CUDA- doesn't exist anymore in the git.
3. Observe the error

**Expected behavior**
Virtual env gets created and 

**Screenshots**
See pastebin mentioned earlier.

**Additional context**
/

**Comment by uselessgames** [comment]
i also have problem #745

**Comment by NoeJuza** [comment]
> i also have problem #745

Not the same error as you, look at the pastebin.

**Comment by tildebyte** [comment]
D'OH

Really sorry about that.

'main' code *should* work with [these reqs](https://pastebin.com/NGNj7qT8) (from 'development' branch) - please let me know here if it doesn't

**Comment by NoeJuza** [comment]
Hello, Thanks for your quick answer, I tried what you recommended (copied the requirements from dev branch (-lin-win-cuda & requirements.txt) but I'm still getting the same exact error (as in the pastebin).

**Comment by tildebyte** [comment]
@NoeJuza;

I downloaded a Windows 11 Developer VM from Microsoft, turned on virtualization on my laptop, fired up the VM (this is the closest I can get to a truly "clean" install), and ran through the install myself. I *did* need to make a couple of tweaks (mostly the repo name thing, but nothing which could cause or fix what you're seeing; also added the correct up-to-date requirements back in), but ultimately I did *not* encounter the issue you're seeing with the pip install.

**Comment by NoeJuza** [comment]
I'll try to reinstall everything from the ground up when I get my machine back, I'll report to you as fast as possible.
Thanks for your work a lot.

**Comment by codefaux** [comment]
I added instructions to #745 which bypass requirements for outdated/unsupported software like pew and/or extras like Miniconda, worth looking into. I'm told my instructions are essentially what pew does, so I don't know why one would use pew instead if it's literally built into Python.

**Comment by tildebyte** [comment]
> set PL_TORCH_DISTRIBUTED_BACKEND=gloo
> ^^ Must be run before running the app.

This is only needed for training

**Comment by codefaux** [comment]
> > set PL_TORCH_DISTRIBUTED_BACKEND=gloo
> > ^^ Must be run before running the app.
> 
> This is only needed for training

Good to know. Since it is needed sometimes maybe it should be enabled by the software? Mentioned somewhere?

**Comment by tildebyte** [comment]
> outdated/unsupported software like pew

pew is neither. Its most actively supported version happens to be in Fedora, currently, but the last release works perfectly well as-is on Windows 10/11 with Python 3.8+

It's just a pure-Python+shell convenience wrapper around standard venvs, so the end-user doesn't have to remember the minute details of every environment they create.

If you really can't stand 'pew', I highly recommend at least looking at [pipenv](https://github.com/pypa/pipenv) as a sane alternative to the default Python venv tools (if you're really curious as to what the big deal is anyway, TAL at (Archive.org) "[Prefer Subshells for Context](https://web.archive.org/web/20201111231240/https://datagrok.org/python/activate/)")

**Comment by tildebyte** [comment]
> > > set PL_TORCH_DISTRIBUTED_BACKEND=gloo
> > > ^^ Must be run before running the app.
> > 
> > 
> > This is only needed for training
> 
> Good to know. Since it is needed sometimes maybe it should be enabled by the software? Mentioned somewhere?

I literally came across it this evening while trying out training for the first time, but yeah needs to be documented.

**Comment by tildebyte** [comment]
BTW, @codefaux, thx for bringing up the pew thing - I'm (re-) familiarizing myself with pipenv, and I'm probably going to implement something using that instead of pew (if only because of its tight integration with pip, and the way it follows "modern" standards so closely)

**Comment by codefaux** [comment]
@tildebyte To prefix; I'm autistic and don't convey tone well. I'm not trying to lecture, just trying to be very precise and clear with my thought process. That said;

I'm unfamiliar with pew so just going off what I heard on the issue itself; someone claimed pew is unsupported, I did not verify that. My bad.

I'm glad you're looking into the situation and considering how to move forward, but I think the act of switching one wrapper for another is moot. I just don't see the point of using a wrapper around an uncomplicated, built-in function that handles itself without dependencies or fuss. A wrapper to simplify a thing when it's super complicated, sure. This -isn't- that, in my view.

Wrappers (such as pew, apparently) are inherently problematic because they; 
- add failure points merely by existing -- evidenced by the individual unable to install due to *something* breaking in a way which provides zero debuggable information. (Here, something probably isn't quoting a path while trying to execute a command.) More moving parts means more breakable parts, and there's no instrumentation for troubleshooting AT ALL in most wrappers.
- obfuscate inner workings, sometimes needlessly (You've informed me that pew installs things to "Windows modern standard" locations, that's convenient but only if you -know- that) without telling the user. Also, what -else- does pew do? If it's just "it provides better default directories" then venv can do that solo with a parameter or two, yeah? Most users are copy-pasting the instructions, they don't care if it has extra words. If it does more, what else is it doing? How do I check that it's doing it right? How do I verify that it's doing things the way I like? Does it -need- to be doing all these things?
- (often) add dependencies, which are just more breakage points in exactly the same ways

What's inherently -wrong- with using just using pip/venv directly?

It isn't even extra commands. It's literally -fewer-, because you don't have to install pew first. It apparently uses the same infrastructure, so I don't see an inherent benefit.

Is it about different default locations? Personally, I LOATHE "Modern Windows standard" style location usage. I'm on an SSD. I don't remember to clean out my random Python caches and site packages and whatever else. I don't know where they are by memory. I DO know where I put the git repo. I DO know that if the venv is in the same place, it'll delete at the same time, and even someone unfamiliar with my computer can attribute the disk usage specifically to the git repo and its venv, at a glance. If I install a repo on my spinning rust and its venv goes there, I don't have to shuffle crap around on my SSD to fit -another- version of pytorch. My personal preferences aside - venv objectively wins for storage management considerations.

How do you activate an environment in a default console with pew, without Python in your path? The venv can be activated using relative directories, after you open the prompt, after you move into the repo directory. Miniconda requires either the shortcut, or remembering to run `Miniconda3\Scripts\Activate.bat <env>` -before- changing dirs in a fresh console, or by typing the entire absolute path name. Even if Python isn't in your PATH variable, venv in the git repo objectively wins for executional convenience, though I admit that's not really a huge perk for many.

What pros do wrappers like pew/pipenv/miniconda provide which are applicable to the end user use case? (Literal question, I'm curious what they actually are.)

TLDR; I don't like wrappers, as a whole, because they cause problems, obfuscate operations, increase dependencies, add a layer of complexity by "handling it for you" -- and sometimes they do it wrong -- literally, thus this conversation.

Do the users know the pros and cons, or do they just want it to work with minimal fuss, and minimal interference outside the repo folder?

**Comment by tildebyte** [comment]
>  I DO know that if the venv is in the same place, it'll delete at the same time

The whole point of virtual environments is that they are DE-coupled from source repos and Python versions (obviously any **specific** venv depends on a **specific** Python).

> How do you activate an environment in a default console with pew, without Python in your path?

Do you seriously NOT have a single system-wide Python in your path? Even if you don't install scripts into system site-packages, you can still use pip's `--user` flag to install, outside of any venv, into your **user** site-packages, to run with system Python.

**Comment by codefaux** [comment]
> The whole point of virtual environments is that they are DE-coupled from source repos and Python versions (obviously any **specific** venv depends on a **specific** Python).
Not if you read the documentation.
```
The [venv](https://docs.python.org/3/library/venv.html#module-venv) module provides support for creating lightweight “virtual environments” with their own site directories, optionally isolated from system site directories. Each virtual environment has its own Python binary (which matches the version of the binary that was used to create this environment) and can have its own independent set of installed Python packages in its site directories.
```
Doesn't say anything about decoupling from source, it specifies the optional capacity to decouple from system site directories. It's MAYBE most commonly used to decouple from source repos, but git allows them to exist within the repo without it syncing so the purpose is still served using them this way. If it was intended to decouple from the working directory tree, it wouldn't default to a relative path when provided nothing but a name. Your understanding sounds like it's based on personal inference or the personal inference of another, not documented intent by the people who made it.

> Do you seriously NOT have a single system-wide Python in your path? Even if you don't install scripts into system site-packages, you can still use pip's `--user` flag to install, outside of any venv, into your **user** site-packages, to run with system Python.
You address my path, but immediately fall back to a usage which places things in user site-packages which is, as I stated, counter to the objective of using venv in the first place.  So to clarify -- users with multiple versions of python installed would perhaps not want one to be defaulted in path to avoid confusion. Python's installer DEFAULTS to NOT add itself to Path, and it's been justified by the team who wrote it. Why are you questioning my accepting the defaults?

Nevermind my accepting the defaults -- why is THAT the point you're trying to address at all?

You have failed to address any of the actual pros to "your way" other than an offhand prior claim about "modern standards" in directory preference. You completely skipped refuting any of the reasoning I put forth, you're just jumping on my Path and misunderstanding the purpose of venv -- the two weakest elements to attack. Okay, so negate that. The whole "path" thing? Skip it. I take that part back, and ironically that removes the tiniest, weakest element of the justifications I've provided.

The thoroughly explained, only VAGUELY DIFFERENT methodology I've suggested produces a working output without involving an extra cog, provided explicitly because your suggested extra cog broke. Your only registered direct argument against excluding the use of an extra cog has been the personal preference of a difference in file location, which is attained by giving venv a different home just as easily as using an extra wrapper that might break. It sounds like you just want to use a wrapper because you want to use a wrapper and that's fine.

**Comment by NoeJuza** [comment]
> I'll try to reinstall everything from the ground up when I get my machine back, I'll report to you as fast as possible. Thanks for your work a lot.

@tildebyte  I reinstalled completely everything from the ground up and now it works, thanks a lot!

**Comment by tildebyte** [comment]
@NoeJuza;

Huzzah! Did you follow the EZ PZ guide, or something else?

**Comment by NoeJuza** [comment]
Just followed brainlessly the Ez pz guide (as it should be IG)

**Comment by tildebyte** [comment]
That's good feedback, thx
